python/dgl/sampling/labor.py

```diff
@@ -0,0 +1,276 @@
+#
+#   Copyright (c) 2022 by Contributors
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+#   Based off of neighbor.py
+#
+
+
+"""Labor sampling APIs"""
+
+from .._ffi.function import _init_api
+from .. import backend as F
+from ..base import DGLError, NID, EID
+from ..heterograph import DGLHeteroGraph
+from .. import ndarray as nd
+from .. import utils
+from .utils import EidExcluder
+from ..backend import from_dgl_nd
+
+__all__ = [
+    'sample_labors']
+
+def sample_labors(g, nodes, fanout, random_seed, edge_dir='in', prob=None,
+                     importance_sampling=0,
+                     copy_ndata=True, copy_edata=True, exclude_edges=None, output_device=None):
+    """Sample neighboring edges of the given nodes and return the induced subgraph.
+
+    For each node, a number of inbound (or outbound when ``edge_dir == 'out'``) edges
+    will be randomly chosen.  The graph returned will then contain all the nodes in the
+    original graph, but only the sampled edges.
+
+    Node/edge features are not preserved. The original IDs of
+    the sampled edges are stored as the `dgl.EID` feature in the returned graph.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The graph.  Can be either on CPU or GPU.
+    nodes : tensor or dict
+        Node IDs to sample neighbors from.
+
+        This argument can take a single ID tensor or a dictionary of node types and ID tensors.
+        If a single tensor is given, the graph must only have one type of nodes.
+    fanout : int or dict[etype, int]
+        The number of edges to be sampled for each node on each edge type.
+
+        This argument can take a single int or a dictionary of edge types and ints.
+        If a single int is given, DGL will sample this number of edges for each node for
+        every edge type.
+
+        If -1 is given for a single edge type, all the neighboring edges with that edge
+        type will be selected.
+    random_seed : tensor with 3 or 4 integers
+        The random seed(s) to be used for sampling along with batch dependency count information
+        stores as the last two elements
```

Comment:
Why do we want to have an individual `random_seed` for LABOR sampler?

Reply:
Because when doing distributed sampling, if seeds are the same, then every machine rolls same random numbers as if everything is running on a single machine. So labor sampling can do distributed sampling with no communication.

Reply:
Cooperative minibatching PR makes use of this feature.

Reply:
One extra use case makes use of having 2 random seeds also.

Reply:
Hmm having this in the `sample_labor` function seems adding additional complexity to the interface.  Is it possible to set the seed outside this function?

Reply:
I store the seed and also the count information in the LaborSampler class and pass it into sample_labors. Where else do you think I can set it and make it so that the c++ side receives it?

Reply:
For instance, `dgl.random.seed()` can set the C++ seeds in DGL.

Reply:
I also don't understand why we need an array of seeds.

Reply:
2 random seeds are used when one wants to use batch_dependency, making consecutive batches dependent on each other by smoothly switching between 2 seeds over batch_dependency batches. The other 2 numbers keep track of how many batches were generated and the batch dependency information. If I use a different seed for each different batch, then the batches are fully independent. If I use the same seed for two different batches, then they are fully dependent. If I want to have something inbetween these two use cases, then I can pass 2 seeds along with how much I want to use the first seed and how much of the second seed. Then it will be a combination of the two seeds and something inbetween having fully independent and dependent batches.

This use case is not in the arxiv paper, it will explained in the cooperative minibatching paper. Also, I didn't want to depend on DGL's seed because if the user uses DGL PRNG for other uses, then it will change the state of the PRNG.

Reply:
Making batches dependent on each other has the advantage that the rolled randoms will be same for same vertices. Then more of the same vertices will be sampled. This reduces input feature requirements across multiple batches.

Reply:
If one uses a gpu cache in the #4341 to cache input features, one will observe that using a larger batch_dependency will improve the cache hit rate.

Reply:
Hmm if that's the case, I would move the `batch_dependency`-related code to the cooperative minibatching PR (or another PR dedicated for cooperative minibatching on LABOR sampling).

Speaking of the interface, I think the core idea is to have a more fine-grained RNG control on each operator.  I don't think supplying seeds in the operators is a good idea as (1) the RNG will then be coupled with the operator itself, (2) it adds extra complexity to the operator's signature, and (3) it conflicts with `dgl.random.seed` which can cause extra confusion to the users.  PyTorch for instance does fine-grained RNG control with context managers, like
```python
with torch.random.fork_rng():
    torch.random.set_rng_state(...)
    x = torch.randn(...)
```
We can borrow the same idea, where you can simply extend the context manager to control how much you would like to mix between the two seeds (if you want to):
```python
rng = dgl.random.cooperative_rng([seed1, seed2], ratio=[0.2, 0.8])
with rng:
    sg = dgl.sampling.sample_labors(g, seeds, fanout, ...)
```

Reply:
These are actually part of the labor sampling algorithm, no other sampling algorithm can do such things as far as I know. However, they are not documented anywhere yet, but will be soon.

I would prefer if it would get in with this functionality as the code review takes time and by the time this code is ready to be merged, I will be closer to completing an arxiv paper on cooperative minibatching. If an explanation is required, I could write an informal post to document this functionality and its benefits and experimental results, and the docstrings can point to the blog so that everything is documented and explained properly. (I already have some preliminary results)

If the above doesn't sound good to you, then since I want the PR to move forward, I will remove the random seeds and batch and layer dependency arguments. In the future, I will open another PR asking to change the interface from top to bottom. Please let me know what you think.

Reply:
> If an explanation is required, I could write an informal post to document this functionality and its benefits and experimental results, and the docstrings can point to the post so that everything is documented and explained properly.

That will be awesome.  I would imagine that it can justify some of the design/implementation decisions (e.g. the necessity of having individual RNG for each vertex, the necessity of using PCG), etc.  An informal post is enough - we just need some place to hold the information on why we are implementing the algorithm in these particular ways.

Reply:
For now, I will hold off on having the multiple random seeds in this PR. For the rest of the design decisions in this PR, I can still write a README in the LABOR sampling example PR. The example PR can also maybe include all of the possible things that LABOR can do, along with NIDs and random seed tricks with code snippets. Since you asked me whether CPU parallelism is possible, it is actually possible to parallelize labor sampling from the python side. I can also explain how this can be done in the README in the example PR.

Reply:
I would prefer to describe the design decision in this PR, since (1) it's harder to lose track, and (2) it's actually different from what LABOR sampling can do; it's more about "why we are implementing LABOR sampling this way".  I don't think the example PR is the best place for having them.  What do you think?

Reply:
Okay then, I will update the docstrings to be as explanative as I can. Will put a few more commends in the C++ implementation explaining how things work.